ctapipe/io/eventfilereader.py

```diff
@@ -148,216 +148,207 @@ def check_file_compatibility(file_path):
-             use_event_id=False):
+        # Check we are within max_events range
+        if not use_event_id and self.max_events and item >= self.max_events:
+            msg = "Event index {} outside of specified max_events {}"\
+                .format(item, self.max_events)
+            raise IndexError(msg)
+
+        return self._get_event(item, use_event_id)
+
+    def _get_event(self, ev, use_event_id):
         """
-        Read the file using the appropriate method depending on the file origin
+        Method for extracting a paritcular event for this data format.
+        If a data format allows random event access, this function can be
+        overrided for a more efficient method.
 
         Parameters
         ----------
-        allowed_tels : list[int]
-            select only a subset of telescope, if None, all are read. This can
-            be used for example emulate the final CTA data format, where there
-            would be 1 telescope per file (whereas in current monte-carlo,
-            they are all interleaved into one file)
-        requested_event : int
-            Seek to a paricular event index
+        ev : int or str
+            The event_index or event_id to seek.
         use_event_id : bool
-            If True ,'requested_event' now seeks for a particular event id
-            instead of index
+            Whether ev is event_id or event_index.
 
         Returns
         -------
-        source : generator
-            A generator that can be iterated over to obtain events
-        """
+        event : ctapipe.io.container
+            The event container filled with the requested event's information
 
-        # Obtain relevent source
-        self.log.debug("Reading file...")
-        if self.max_events:
-            self.log.info("Max events being read = {}".format(self.max_events))
-        source = hessio_event_source(self.input_path,
-                                     max_events=self.max_events,
-                                     allowed_tels=allowed_tels,
-                                     requested_event=requested_event,
-                                     use_event_id=use_event_id)
-        self.log.debug("File reading complete")
-        return source
+        """
+        if not use_event_id:
+            msg = "Event index {} not found in file".format(ev)
+            for event in self.source:
+                if event.count == ev:
+                    self.current_event = event
+                    return deepcopy(event)
+        else:
+            msg = "Event id {} not found in file".format(ev)
+            for event in self:  # Event Ids may not be in order
+                if event.r0.event_id == ev:
+                    self.current_event = event
+                    return deepcopy(event)
+        raise IndexError(msg)
+
+    def __len__(self):
+        if self.is_stream:
+            raise IOError("Input is a stream, __len__ is disabled")
```

Comment:
This is the exact same argument you could use against removing `__getitem__` from the base class and forcing the sub-classes to define it. However in this case with the streams, it is physically impossible (without some extra functionality like caching). 

As it it only makes sense to use streams when using the pipeline processing, where we go through event by event, I don't think we should be concerned that this throws an error if someone tries to use a stream in their debugging script or GUI, they should clearly be using the file itself.

Reply:
> This is the exact same argument you could use against removing __getitem__ 
We did.

We see here maybe another hierarchy would be appropriate:
`
class EventReader:
    pass

class EventFileReader(EventReader):
    pass

class EventStreamReader(EventReader):
    pass
````

Reply:
Yes...maybe another hierarchy... but the one you've suggested would require defining how the events are read for each data format twice.

How about a wrapping class (not sure if theres a python convention on how to do these):

```
class EventFileReader: 
   pass

class EventSeeker:
   def __init__(eventfilereader):
      pass

   def __getitem___(self):
      pass
```

EventFileReader is exactly as we have it, but without __getitem__. Each data format creates their own Reader that inherits this and defines how the events containers are filled

EventSeeker takes an EventFileReader, and defines __getitem__ as a loop through the EventFileReader (as it currenly exists in EventFileReader.__getitem__), unless that particular EventFileReader defines __getitem__ itself (because that data format has a random event retrieval method).

How does that sound?

Reply:
Then in that case EventFileReader is only used to loop through events, compatible for both files and streams, and is @dneise and your's desire to have a more basic EventFileReader definition. Whereas EventSeeker is defined for those who want `__getitem__` functionality and takes any EventFileReader as input.